### DISCUSSION TITLE
VanX 0.6.0: `vanX.noreactive` - exemption from reactivity conversion

### DISCUSSION THREAD

**Discussion by Tao-VanJS** [vanjs-org/van#311] [Announcements]
Hi fellow VanJSers,

I'm happy to announce the release of **VanX** `0.6.0`. 🎉🎉🎉

This release adds a new top-level function `vanX.noreactive`. `vanX.noreactive` is used while constructing the reactive objects to exempt certain fields from being converted into reactive objects. For instance, if we have an application with a reactive array like:

```js
const data = vanX.reactive([
  vanX.noreactive(new ArrayBuffer(8)),
  vanX.noreactive(new ArrayBuffer(16)),
])

const App = () => div(
  vanX.list(div, data, v => div(v.val.byteLength)),
  div(button({onclick: () => data.push(vanX.noreactive(new ArrayBuffer(24)))}, "Push")),
)
```

The `ArrayBuffer`s in `data` will be treated as primitive fields without being converted into reactive objects. This feature is essential as the objects of certain native or 3rd party classes can't be correctly converted into reactive objects. `ArrayBuffer` is one of such classes as wrapping it around a Proxy will cause problems (e.g.: #310).

---

With this new function, the bundle size of **VanX** increases slightly. Gzipped bundle increases to `1150 bytes` (1.1kB) from `1130 bytes` (1.1kB) (`20 bytes` increase), while minified bundle increases to `2197 bytes` (2.1kB) from `2150 bytes` (2.1kB) (`47 bytes` increase) - still being an ultra-lightweight extension to **VanJS**.

❤️ Hope you can enjoy!

**Comment by HEAVYPOLY** [comment]
Awesome, thank you!  Maybe a future feature, would it make sense if this was auto detected/applied if the field is not able to be reactive? 
Maybe vanX.replace could do the detection?

**Reply by Tao-VanJS** [reply]
That will be hard. Weather the objects of a specific class can be turned into reactive or not depends on the implementation details of the class.

**Reply by HEAVYPOLY** [reply]
hmm...what about an option that specifies how many levels of reactiveness to go?

Like vanX.replace(this, that, 2). so that everything after 2 nested levels becomes vanX.noreactive.

Maybe not a great solution actually...I'm trying to think of a way objects like these could work automatically
```js
[
  {
    a: 'string',
    b: 7,
    c: new ArrayBuffer(16),
  },
  {
    a: 'blah',
    b: 20,
    c: new ArrayBuffer(32),
  }
]
```

**Comment by Tao-VanJS** [comment]
**VanX** `0.6.1` was just released. **VanX** `0.6.1` fixed the bug #356.

Size change:
* Gzipped bundle: `1150 bytes` (1.1kB) to `1187 bytes` (1.2kB)
* Minified bundle `2150 bytes` (2.1kB) to `2267 bytes` (2.2kB)

**Comment by Tao-VanJS** [comment]
**VanX** `0.6.2` was just released. **VanX** `0.6.1` changed the `type` in `package.json` to `module`, which enables better integrations with frameworks like Vite.

There is no change to the actual implementation.

**Comment by Tao-VanJS** [comment]
**VanX** `0.6.3` was just released. **VanX** `0.6.3` fixed the bug #444, where the prototypes of reactive objects are not kept in `vanX.compact`.

Size change:
* Gzipped bundle: `1187 bytes` (1.2kB) to `1210 bytes` (1.2kB)
* Minified bundle: `2267 bytes` (2.2kB) to `2296 bytes` (2.2kB)

Thanks @reserved-word for reporting the bug!
